homeassistant/components/tesla_fleet/climate.py

```diff
@@ -0,0 +1,300 @@
+"""Climate platform for Tesla Fleet integration."""
+
+from __future__ import annotations
+
+from itertools import chain
+from typing import Any, cast
+
+from tesla_fleet_api.const import CabinOverheatProtectionTemp, Scope
+
+from homeassistant.components.climate import (
+    ATTR_HVAC_MODE,
+    ClimateEntity,
+    ClimateEntityFeature,
+    HVACMode,
+)
+from homeassistant.const import (
+    ATTR_TEMPERATURE,
+    PRECISION_HALVES,
+    PRECISION_WHOLE,
+    UnitOfTemperature,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import TeslaFleetConfigEntry
+from .const import DOMAIN, TeslaFleetClimateSide
+from .entity import TeslaFleetVehicleEntity
+from .helpers import handle_vehicle_command
+from .models import TeslaFleetVehicleData
+
+DEFAULT_MIN_TEMP = 15
+DEFAULT_MAX_TEMP = 28
+
+PARALLEL_UPDATES = 0
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: TeslaFleetConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Tesla Fleet Climate platform from a config entry."""
+
+    async_add_entities(
+        chain(
+            (
+                TeslaFleetClimateEntity(
+                    vehicle, TeslaFleetClimateSide.DRIVER, entry.runtime_data.scopes
+                )
+                for vehicle in entry.runtime_data.vehicles
+            ),
+            (
+                TeslaFleetCabinOverheatProtectionEntity(
+                    vehicle, entry.runtime_data.scopes
+                )
+                for vehicle in entry.runtime_data.vehicles
+            ),
+        )
+    )
+
+
+class TeslaFleetClimateEntity(TeslaFleetVehicleEntity, ClimateEntity):
+    """Tesla Fleet vehicle climate entity."""
+
+    _attr_precision = PRECISION_HALVES
+
+    _attr_temperature_unit = UnitOfTemperature.CELSIUS
+    _attr_hvac_modes = [HVACMode.HEAT_COOL, HVACMode.OFF]
+    _attr_supported_features = (
+        ClimateEntityFeature.TURN_ON
+        | ClimateEntityFeature.TURN_OFF
+        | ClimateEntityFeature.TARGET_TEMPERATURE
+        | ClimateEntityFeature.PRESET_MODE
+    )
+    _attr_preset_modes = ["off", "keep", "dog", "camp"]
+    _enable_turn_on_off_backwards_compatibility = False
+
+    def __init__(
+        self,
+        data: TeslaFleetVehicleData,
+        side: TeslaFleetClimateSide,
+        scopes: Scope,
+    ) -> None:
+        """Initialize the climate."""
+        self.scoped = Scope.VEHICLE_CMDS in scopes
+        if not self.scoped or data.signing:
```

Comment:
There are not sensors for a read only view because they would duplicate the values exposed by the climate entity.

The intent is to use the correct platform so that a user can later decide to add the required scopes and the entity won't have to change.

The approach used here is consistent with Tessie and Teslemetry.

Reply:
There's climate_state_inside_temp which seemed like the important bit.

Either way, my other point was there is complexity with these checks later in the entity and I don't think it is needed. My impression is that if the features aren't set the code in the entity for double checking these things won't run because of the missing featuers. My assumption is the tests exercising those aren't actually covering those code paths (could confirm looking at code coverage or the error messages raised).

Reply:
Some of the climate attributes are exposed as sensors yes (forgot I did that), but this isnt true for the other platforms for example the lock platform is the only way to see if the vehicle is locked. I dont believe hiding the climate entities is the right outcome, because I know for sure I'll have users raising issues that the climate entity is missing.

These cases are covered in tests `test_climate_noscope` and `test_climate_command_signing`.

I dont think I understand your core point because there is no "double checking". If there is a reason the integration knows a service call will fail it checks for that before trying to execute it, and only does that at run time.

Reply:
Ok I do think I get what you mean by "double checking" now. My challenge is that not all service calls can be disabled with support_features.

Is your suggestion to remove the `self.raise_for_signing()` and `self.raise_for_scope()` on service calls that are removed with `ClimateEntityFeature(0)`?

Reply:
Ok so you can't disable set HVAC mode since it's not associated with a feature? Presumably because it's a requirement to make a useful climate entity. There is lots of prior art of not creating dummy entities when the user doesn't allow the scopes too fwiw (e.g. other integrations use oauth scopes to decide what to expose).

Perhaps simplify to "read only" since the details of why it's read only (scopes, code signing, etc) doesn't matter and is set only once.

Reply:
> other integrations use oauth scopes to decide what to expose
Do you have an example of what good looks like? Am I meant to created disabled by default sensor entities to cover every single other platform type just incase the user doesnt have scopes? I obviously cannot remove sensor entities based on scopes because that would be a breaking change when scopes change.

> Perhaps simplify to "read only"
Not all commands require signing, so I only `raise_for_signing` on the commands that do.

Reply:
> Do you have an example of what good looks like? 
The ones i've had experience with (Nest, Fitbit, Google Calendar, etc) each follow the user choice during oauth / setup on what permissions are allowed then only expose entities based on that choice. They don't expose placeholder functionality.

I do not intend to suggest exposing disabled by default sensor entities -- saying the opposite, about only exposing entities based on what was allowed.

I did see knx appears to have a notion of a read only climate entity. I noticed it returns a single `hvac_mode` for `hvac_modes`. That may have the side effect that it doesn't allow you to change the mode?

> Not all commands require signing

Which commands require only signing or only scopes? The logic looked the same everywhere in climate entity so it didn't seem necessary to be specific and check both.

Reply:
Interesting idea about `hvac_modes` ill give that a shot.

In the climate platform everything requires signing, but there are commands in other platforms that dont, however looking at the list its rather small, and when checking the Tesla documentation they now suggest signing for all commands (despite some not being implemented in their own reference implementation), so I can probably remove this distinction.

Reply:
Ah -- i think i was missing that you're adding this to the parent class and assuming some future expansion. My take would be to just do what climate needs for this PR and consider other uses later.  This may be why i'm saying its complex because i'm looking at the needs for climate which seem pretty modest/minimal.

Reply:
Its one of the downsides of the 1 PR per platform approch, you dont get to see the full context I am building towards. I did climate first because it touches most functionality but only has two entities so its smaller.

I have taken some of your suggestions and now implement a `self.read_only` property and am using that instead. I still want to keep raise_for_scope and raise_for_signing seperate for maintainability reasons.

Do you still feel I need to remove the climate entity all together, or does the modified HVAC_MODES with the zero supported features suitable?

Reply:
I usually think what works is to add exactly what is needed in each PR for just that platform and make it more general one step at a time.  

I can see that representing this as a bunch of separate objects would be more complex than a climate entity, so i'm open to keeping it as is.

Reply:
> more general one step at a time.
The problem is, 95% of the integration has already been written as is just being copied from already merged code, so doing that would actually require me to re-write it in a more specific fashion, only to slowly rework it back to being generic again in subsequent PRs, and hopefully end up where I started. This first PR is always the hardest, but then since its layed the groundwork I'll be able to raise the 6 or 7 other platform PRs without merge conflicts.

Reply:
I don't recommend writing an entire integration with many platforms up front without feedback on the first platform.

Reply:
But I **_did_** get feedback, and they are platinum quality.

https://github.com/home-assistant/core/tree/dev/homeassistant/components/teslemetry
https://github.com/home-assistant/core/tree/dev/homeassistant/components/tessie

Reply:
Oh apologies, I didn't get part of your point before. To clarify I didn't intent to comment on the quality. I'm saying that sending PRs and being open to feedback avoids "95% of the integration has already been written as is just being copied from already merged code, so doing that would actually require me to re-write it in a more specific fashion".  I was making a more general comment about being open to things changing after initial platform.

I understanding now you're saying you are copying the code from an existing integration and so you have different expectations.

Reply:
Please dont misunderstand, I am also not trying to comment on the quality of my code, while it would be nice if this PR could have slide through without changes because most of it had already been reviewed, you have helped improve the quality of this and my other integrations.

I have plenty of feelings about the Home Assistant PR rules and how they making contributing harder, but i know the point is to make the reviews easier because that is actually the bottleneck. Moving forward I'll reference the Teslemetry integration file so reviewers can have context about the bigger picture and what I am copy pasting from.